### PR TITLE
Return count of repos that fail to clone or fetch

### PR DESCRIPTION
**Pull request by ryanslade**
With this change we now return a message specifying the number of repos that
failed to be cloned or fetched in the status message endpoint so that it can be
displayed in the notifications area.

Closes: https://github.com/sourcegraph/sourcegraph/issues/19755

### CODE REVIEW

File: internal/database/repos.go [hunk 1]
```diff
@@ -714,6 +719,9 @@ func (s *RepoStore) listSQL(ctx context.Context, opt ReposListOptions) (*sqlf.Qu
 		// where = append(where, sqlf.Sprintf("gr.clone_status = 'cloned'"))
 		where = append(where, sqlf.Sprintf("(gr.clone_status = 'cloned' OR (gr.clone_status IS NULL AND repo.cloned))"))
 	}
+	if opt.FailedSync {
+		where = append(where, sqlf.Sprintf("gr.last_error IS NOT NULL"))
```
Comment (tsenart) [inline internal/database/repos.go:723]: What's the query plan in the production database? Perhaps we'll need an index.

File: internal/database/repos.go [hunk 2]
```diff
@@ -469,6 +469,11 @@ type ReposListOptions struct {
 	// is non-zero. Note that these are not repos owned by this user, just ones they are interested in.
 	IncludeUserPublicRepos bool
 
+	// FailedSync, if true, will filter to only repos that failed to sync when last
+	// attempted. Specifically, this means that they have a non-null last_error value
+	// in the gitserver_repos table.
+	FailedSync bool
```
Comment (tsenart) [inline internal/database/repos.go:475]: Sync is usually associated with external service syncing right? Would a better name be `FailedGitFetching`?

Comment (ryanslade) [reply internal/database/repos.go]: I called it `FailedFetch` since we don't only support git now :)

File: internal/database/repos.go (hunk 1 above)
Comment (ryanslade) [reply internal/database/repos.go:723]: Currently we only combine this query with a list of external services so the query plan is pretty good, for example:

```
Aggregate  (cost=770.99..771.00 rows=1 width=8) (actual time=0.037..0.038 rows=1 loops=1)
  ->  Nested Loop  (cost=1.30..770.98 rows=1 width=0) (actual time=0.035..0.036 rows=0 loops=1)
        ->  Nested Loop  (cost=0.86..594.52 rows=379 width=8) (actual time=0.023..0.025 rows=1 loops=1)
              ->  Index Only Scan using external_service_repos_idx on external_service_repos esr  (cost=0.43..9.26 rows=379 width=4) (actual time=0.010..0.011 rows=1 loops=1)
                    Index Cond: (external_service_id = 120)
                    Heap Fetches: 1
              ->  Index Only Scan using repo_pkey on repo  (cost=0.43..1.54 rows=1 width=4) (actual time=0.011..0.011 rows=1 loops=1)
                    Index Cond: (id = esr.repo_id)
                    Heap Fetches: 1
        ->  Index Scan using gitserver_repos_pkey on gitserver_repos gr  (cost=0.43..0.47 rows=1 width=4) (actual time=0.010..0.011 rows=0 loops=1)
              Index Cond: (repo_id = repo.id)
              Filter: (last_error IS NOT NULL)
              Rows Removed by Filter: 1
Planning Time: 0.522 ms
Execution Time: 0.072 ms

```

But, we don't know how someone might combine it in future so I think to be safe we should add an index where `last_error IS NOT NULL`

**Comment by ryanslade** [comment]
@tsenart PTAL

**Review by tsenart — APPROVED** [review]

**Pull request merged**